api4/system.go

```diff
@@ -444,3 +449,49 @@ func pushNotificationAck(c *Context, w http.ResponseWriter, r *http.Request) {
 
 	ReturnStatusOK(w)
 }
+
+func setServerBusy(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !c.App.SessionHasPermissionTo(c.App.Session, model.PERMISSION_MANAGE_SYSTEM) {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	// number of seconds to keep server marked busy
+	secs := r.URL.Query().Get("secs")
+	if secs == "" {
+		secs = "3600" // default 1 hour
+	}
+
+	i, err := strconv.ParseInt(secs, 10, 64)
+	if err != nil || i <= 0 {
+		c.SetInvalidUrlParam("secs")
+		return
+	}
+
+	c.App.Srv.Busy.Set(time.Second * time.Duration(i))
+	mlog.Warn("server busy state activitated - non-critical services disabled", mlog.Int64("seconds", i))
+	ReturnStatusOK(w)
+}
+
+func clearServerBusy(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !c.App.SessionHasPermissionTo(c.App.Session, model.PERMISSION_MANAGE_SYSTEM) {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+	c.App.Srv.Busy.Clear()
+	mlog.Info("server busy state cleared - non-critical services enabled")
+	ReturnStatusOK(w)
+}
+
+func getServerBusyExpires(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !c.App.SessionHasPermissionTo(c.App.Session, model.PERMISSION_MANAGE_SYSTEM) {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	m := make(map[string]string)
+	m["busy"] = fmt.Sprintf("%t", c.App.Srv.Busy.IsBusy())
+	m["expires"] = fmt.Sprintf("%d", c.App.Srv.Busy.Expires().Unix())
+	m["expires_ts"] = c.App.Srv.Busy.Expires().UTC().Format("Mon Jan 2 15:04:05 -0700 MST 2006")
```

Comment:
What about making a struct for this? That way the client (in model/client4.go) can decode into that and it's clear to API users what the response structure is

Reply:
Good idea.  Done via https://github.com/mattermost/mattermost-server/pull/13212/commits/d3706fb9ecb362c9d539d373c0f5b2fa147e2442